### PR TITLE
Implement `cubic_interpolate()` as MathFunc for refactoring

### PR DESCRIPTION
**Pull request by TokageItLab**
Implement `cubic_interpolate()` as built-in math function for refactoring animation and anywhere.

This is the first step in solving #57952. We need to change the formula to solve #57952 completely, but since there were many duplicated formulas, this change put them in one place.

### CODE REVIEW

File: editor/animation_track_editor.cpp [hunk 1]
```diff
@@ -4952,7 +4952,7 @@ void AnimationTrackEditor::_add_method_key(const String &p_method) {
 	EditorNode::get_singleton()->show_warning(TTR("Method not found in object: ") + p_method);
 }
 
-void AnimationTrackEditor::_key_selected(int p_track, int p_key, bool p_single) {
+void AnimationTrackEditor::_key_selected(int p_key, bool p_single, int p_track) {
```
Comment (akien-mga) [inline editor/animation_track_editor.cpp:4955]: Already fixed by #57946.

File: modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs [hunk 2]
```diff
@@ -179,6 +179,25 @@ public static real_t Cosh(real_t s)
             return (real_t)Math.Cosh(s);
         }
 
+        /// <summary>
+        /// Linearly interpolates between two values by a normalized value.
+        /// This is the opposite <see cref="InverseLerp(real_t, real_t, real_t)"/>.
```
Comment (akien-mga) [inline modules/mono/glue/GodotSharp/GodotSharp/Core/Mathf.cs:184]: That's a copy paste from the lerp docs, needs to be changed.

File: core/math/math_funcs.h [hunk 3]
```diff
@@ -235,6 +235,21 @@ class Math {
...
+						(-p_pre + p_to) * p_weight +
+						(2.0 * p_pre - 5.0 * p_from + 4.0 * p_to - p_post) * p_weight * p_weight +
+						(-p_pre + 3.0 * p_from - 3.0 * p_to + p_post) * p_weight * p_weight * p_weight);
+	}
+	static _ALWAYS_INLINE_ float cubic_interpolate(float p_from, float p_to, float p_pre, float p_post, float p_weight) {
+		return 0.5f *
+				((p_from * 2.0f) +
+						(-p_pre + p_to) * p_weight +
+						(2.0f * p_pre - 5.0f * p_from + 4.0f * p_to - p_post) * p_weight * p_weight +
+						(-p_pre + 3.0f * p_from - 3.0f * p_to + p_post) * p_weight * p_weight * p_weight);
```
Comment (aaronfranke) [inline core/math/math_funcs.h:250]: It's hard to say for sure without digging around in Godbolt, but I think the compiler can optimize this more if we group together the `p_weight` multiplications:

```suggestion
						(2.0f * p_pre - 5.0f * p_from + 4.0f * p_to - p_post) * (p_weight * p_weight) +
						(-p_pre + 3.0f * p_from - 3.0f * p_to + p_post) * (p_weight * p_weight * p_weight));
```

**Comment by sairam4123** [comment]
I'd recommend calling it cubic_lerp instead to be in par with lerp function.

**Comment by aaronfranke** [comment]
@sairam4123 The "l" in "lerp" stands for "linear" so "cubic_lerp" would not be an appropriate name. We discussed the naming back when "lerp" was called "linear_interpolate" and decided to shorten it to "lerp", but "cubic_interpolate" does not have a common shortened version, "cerp" would sounds weird.

**Comment by TokageItLab** [comment]
As aaronfranke said, lerp is itself an acronym for linear-interpolate. If you want to make a comparison, it would be more correct to make lerp() into linear_interpolate(), but I don't think it is worth it.

**Comment by sairam4123** [comment]
But there is slerp, so I thought cubic_lerp could work.On 13-Feb-2022 1:12 PM, Aaron Franke ***@***.***> wrote:
@sairam4123 The "l" in "lerp" stands for "linear" so "cubic_lerp" would not be an appropriate name. We discussed the naming back when "lerp" was called "linear_interpolate" and decided to shorten it to "lerp", but "cubic_interpolate" does not have a common shortened version, "cerp" would sounds weird.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**